cpp/src/io/json/read_json.cu

```diff
@@ -278,25 +306,51 @@ table_with_metadata read_json(host_span<std::unique_ptr<datasource>> sources,
+  std::for_each(sources.begin(), sources.end(), [](const auto& source) {
+    CUDF_EXPECTS(source->size() < std::numeric_limits<int>::max(),
+                 "The size of each source file must be less than INT_MAX bytes");
+  });
 
-  // If input JSON buffer has single quotes and option to normalize single quotes is enabled,
-  // invoke pre-processing FST
-  if (reader_opts.is_enabled_normalize_single_quotes()) {
-    normalize_single_quotes(bufview, stream, rmm::mr::get_current_device_resource());
-  }
+  size_t const batch_size = std::numeric_limits<int>::max();
+  int const num_batches   = std::ceil((double)sources_size(sources, 0, 0) / batch_size);
+  if (num_batches == 1) return read_batch(sources, reader_opts, stream, mr);
 
-  // If input JSON buffer has unquoted spaces and tabs and option to normalize whitespaces is
-  // enabled, invoke pre-processing FST
-  if (reader_opts.is_enabled_normalize_whitespace()) {
-    normalize_whitespace(bufview, stream, rmm::mr::get_current_device_resource());
-  }
+  std::vector<size_t> prefsum_source_sizes(sources.size());
+  std::transform_inclusive_scan(sources.begin(),
+                                sources.end(),
+                                prefsum_source_sizes.begin(),
+                                std::plus<size_t>{},
+                                [](const std::unique_ptr<datasource>& s) { return s->size(); });
 
-  auto buffer =
-    cudf::device_span<char const>(reinterpret_cast<char const*>(bufview.data()), bufview.size());
-  stream.synchronize();
-  return device_parse_nested_json(buffer, reader_opts, stream, mr);
+  auto start_source_it = sources.begin();
+  std::vector<cudf::io::table_with_metadata> partial_tables;
+  for (int batch = 0; batch < num_batches; batch++) {
+    auto end_prefsum_it = std::upper_bound(
+      prefsum_source_sizes.begin(), prefsum_source_sizes.end(), (batch + 1) * batch_size);
+    auto end_source_it =
+      sources.begin() + std::distance(prefsum_source_sizes.begin(), end_prefsum_it);
+    partial_tables.emplace_back(
+      read_batch(host_span<std::unique_ptr<datasource>>(
+                   start_source_it, std::distance(start_source_it, end_source_it)),
+                 reader_opts,
+                 stream,
+                 mr));
+    start_source_it = end_source_it;
+  }
+  auto partial_table_views = std::vector<cudf::table_view>(partial_tables.size());
+  std::transform(
+    partial_tables.begin(), partial_tables.end(), partial_table_views.begin(), [](auto& table) {
+      return table.tbl->view();
+    });
+  std::vector<column_name_info> concatenated_table_schema_info;
+  std::for_each(
+    partial_tables.begin(), partial_tables.end(), [&concatenated_table_schema_info](auto& ptbl) {
+      auto const& ptbl_schema = ptbl.metadata.schema_info;
+      concatenated_table_schema_info.insert(
+        concatenated_table_schema_info.end(), ptbl_schema.begin(), ptbl_schema.end());
+    });
```

Comment:
this is surprising. doesn't schema_info store per-column information?

Reply:
The current plan is to fail in case the schema info does not match across all batches.